Allow disabling int specialization on nn.Modules

jhadidjojo opened the pull request:
Resolves issue #140464 by adding an option to not specialize int from nn.Modules (False by default to maintain existing behavior).

Test Plan: `buck2 test mode/opt caffe2/test/dynamo:test_dynamo -- test_modules.py::NNModuleTests::test_nn_module_unspec_int_attr`

Differential Revision: D66837042




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jhadidjojo commented:
@pytorchbot label "release notes: dynamo"

jhadidjojo commented:
@pytorchbot merge

yanboliang commented on torch/_dynamo/config.py:
```diff
@@ -256,6 +256,14 @@
+# Setting this flag to True will allow int members to be potentially unspecialized
+# through dynamic shape mechanism.
+# Defaults to False for BC.
+disable_force_spec_int_on_specialized_nn_module = False
```
`unspec_int_on_nn_module` is a more simple name.

jhadidjojo committed:
Allow disabling int specialization on nn.Modules (#142829)

Summary:
Resolves issue [#140464](https://github.com/pytorch/pytorch/issues/140464) by adding an option to not specialize int from nn.Modules (False by default to maintain existing behavior).


Test Plan:
`buck2 test mode/opt caffe2/test/dynamo:test_dynamo -- test_modules.py::NNModuleTests::test_nn_module_unspec_int_attr`

Also stepped through the repro in [#140464](https://github.com/pytorch/pytorch/issues/140464) to confirm that, with the flag set to True, the int is no longer specialized.

Reviewed By: ezyang, yanboliang

Differential Revision: D66837042

jhadidjojo commented:
@pytorchbot merge

jhadidjojo commented:
@pytorchbot merge -f "Failures unrelated"
